src/main/java/org/betonquest/betonquest/modules/schedule/impl/ExecutorServiceScheduler.java

```diff
@@ -0,0 +1,127 @@
+package org.betonquest.betonquest.modules.schedule.impl;
+
+import lombok.CustomLog;
+import org.betonquest.betonquest.BetonQuest;
+import org.betonquest.betonquest.api.schedule.CatchupStrategy;
+import org.betonquest.betonquest.api.schedule.Schedule;
+import org.betonquest.betonquest.api.schedule.Scheduler;
+import org.betonquest.betonquest.modules.schedule.impl.simple.SimpleScheduler;
+import org.bukkit.Bukkit;
+
+import java.util.Locale;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * A scheduler that already provides a {@link ScheduledExecutorService} for scheduling events to run at a specific
+ * point in (real) time.
+ * Starting and stopping the executor as well as adding/removing schedules is already implemented.
+ *
+ * @param <S> Type of Schedule
+ */
+@SuppressWarnings("PMD.DoNotUseThreads")
+@CustomLog(topic = "Schedules")
+public abstract class ExecutorServiceScheduler<S extends Schedule> extends Scheduler<S> {
+
+    /**
+     * Maximum time that the scheduler will wait on shutdown/reload for currently executing schedules.
+     */
+    private static final int TERMINATION_TIMEOUT_MS = 5;
+
+    /**
+     * BetonQuest instance
+     */
+    protected final BetonQuest betonQuestInstance;
```

Comment:
You should use the `Plugin` type here, because you dont need methods from BetonQuest here